// @datadog/datadog-ci/dist/commands/commit/renderer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderCommandInfo = exports.renderSuccessfulCommand = exports.renderUpload = exports.renderRetriedUpload = exports.renderFailedUpload = exports.renderConfigurationError = exports.renderGitError = void 0;
const chalk_1 = __importDefault(require("chalk"));
const formatting_1 = require("../../helpers/formatting");
const renderGitError = (errorMessage) => chalk_1.default.red(`${formatting_1.ICONS.FAILED} An error occured while invoking git: ${errorMessage}
Make sure the command is running within your git repository.\n`);
exports.renderGitError = renderGitError;
const renderConfigurationError = (error) => chalk_1.default.red(`${formatting_1.ICONS.FAILED} Configuration error: ${error}.\n`);
exports.renderConfigurationError = renderConfigurationError;
const renderFailedUpload = (errorMessage) => chalk_1.default.red(`${formatting_1.ICONS.FAILED} Failed upload: ${errorMessage}\n`);
exports.renderFailedUpload = renderFailedUpload;
const renderRetriedUpload = (errorMessage, attempt) => chalk_1.default.yellow(`[attempt ${attempt}] Retrying upload: ${errorMessage}\n`);
exports.renderRetriedUpload = renderRetriedUpload;
exports.renderUpload = 'Uploading\n';
const renderSuccessfulCommand = (duration, dryRun) => {
    if (dryRun) {
        return chalk_1.default.green(`${formatting_1.ICONS.SUCCESS} [DRYRUN] Handled in ${duration} seconds.\n`);
    }
    else {
        return chalk_1.default.green(`${formatting_1.ICONS.SUCCESS} Uploaded in ${duration} seconds.\n`);
    }
};
exports.renderSuccessfulCommand = renderSuccessfulCommand;
const renderCommandInfo = (dryRun) => {
    if (dryRun) {
        return chalk_1.default.yellow(`${formatting_1.ICONS.WARNING} DRY-RUN MODE ENABLED. WILL NOT UPLOAD\n`);
    }
    return chalk_1.default.green('Starting upload. \n');
};
exports.renderCommandInfo = renderCommandInfo;
