// @datadog/datadog-ci/dist/commands/lambda/__tests__/function.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
jest.mock('../loggroup');
const constants_1 = require("../constants");
const function_1 = require("../function");
const loggroup = __importStar(require("../loggroup"));
const makeMockLambda = (functionConfigs) => ({
    getFunction: jest.fn().mockImplementation(({ FunctionName }) => ({
        promise: () => Promise.resolve({ Configuration: functionConfigs[FunctionName] }),
    })),
    listFunctions: jest.fn().mockImplementation(() => ({
        promise: () => Promise.resolve({ Functions: Object.values(functionConfigs) }),
    })),
    listTags: jest.fn().mockImplementation(() => ({ promise: () => Promise.resolve({ Tags: {} }) })),
    tagResource: jest.fn().mockImplementation(() => ({ promise: () => Promise.resolve() })),
    updateFunctionConfiguration: jest.fn().mockImplementation(() => ({ promise: () => Promise.resolve() })),
});
const makeMockCloudWatchLogs = () => ({});
const mockAwsAccount = '123456789012';
describe('function', () => {
    describe('getLambdaConfigs', () => {
        const OLD_ENV = process.env;
        beforeEach(() => {
            jest.resetModules();
            process.env = {};
        });
        afterAll(() => {
            process.env = OLD_ENV;
        });
        test('returns the update request for each function', () => __awaiter(void 0, void 0, void 0, function* () {
            const lambda = makeMockLambda({
                'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument': {
                    FunctionArn: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
                    Handler: 'index.handler',
                    Runtime: 'nodejs12.x',
                },
            });
            const cloudWatch = makeMockCloudWatchLogs();
            const settings = {
                environment: 'staging',
                flushMetricsToLogs: false,
                layerVersion: 22,
                logLevel: 'debug',
                mergeXrayTraces: false,
                service: 'middletier',
                tracingEnabled: false,
                version: '0.2',
            };
            const result = yield function_1.getLambdaConfigs(lambda, cloudWatch, 'us-east-1', ['arn:aws:lambda:us-east-1:000000000000:function:autoinstrument'], settings);
            expect(result.length).toEqual(1);
            expect(result[0].updateRequest).toMatchInlineSnapshot(`
        Object {
          "Environment": Object {
            "Variables": Object {
              "DD_ENV": "staging",
              "DD_FLUSH_TO_LOG": "false",
              "DD_LAMBDA_HANDLER": "index.handler",
              "DD_LOG_LEVEL": "debug",
              "DD_MERGE_XRAY_TRACES": "false",
              "DD_SERVICE": "middletier",
              "DD_SITE": "datadoghq.com",
              "DD_TRACE_ENABLED": "false",
              "DD_VERSION": "0.2",
            },
          },
          "FunctionName": "arn:aws:lambda:us-east-1:000000000000:function:autoinstrument",
          "Handler": "/opt/nodejs/node_modules/datadog-lambda-js/handler.handler",
          "Layers": Array [
            "arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node12-x:22",
          ],
        }
      `);
        }));
        test('returns results for multiple functions', () => __awaiter(void 0, void 0, void 0, function* () {
            const lambda = makeMockLambda({
                'arn:aws:lambda:us-east-1:000000000000:function:another-func': {
                    FunctionArn: 'arn:aws:lambda:us-east-1:000000000000:function:another-func',
                    Runtime: 'nodejs12.x',
                },
                'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument': {
                    Environment: {
                        Variables: {
                            [constants_1.FLUSH_TO_LOG_ENV_VAR]: 'false',
                            [constants_1.LAMBDA_HANDLER_ENV_VAR]: 'index.handler',
                            [constants_1.LOG_LEVEL_ENV_VAR]: 'debug',
                            [constants_1.MERGE_XRAY_TRACES_ENV_VAR]: 'false',
                            [constants_1.SITE_ENV_VAR]: 'datadoghq.com',
                            [constants_1.TRACE_ENABLED_ENV_VAR]: 'false',
                            [constants_1.SERVICE_ENV_VAR]: 'middletier',
                            [constants_1.ENVIRONMENT_ENV_VAR]: 'staging',
                            [constants_1.VERSION_ENV_VAR]: '0.2',
                        },
                    },
                    FunctionArn: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
                    Runtime: 'nodejs12.x',
                },
            });
            const cloudWatch = makeMockCloudWatchLogs();
            const settings = {
                environment: 'staging',
                flushMetricsToLogs: false,
                layerVersion: 23,
                mergeXrayTraces: false,
                service: 'middletier',
                tracingEnabled: false,
                version: '0.2',
            };
            const result = yield function_1.getLambdaConfigs(lambda, cloudWatch, 'us-east-1', [
                'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
                'arn:aws:lambda:us-east-1:000000000000:function:another-func',
            ], settings);
            expect(result.length).toEqual(2);
        }));
    });
    describe('getFunctionConfiguration', () => {
        const OLD_ENV = process.env;
        beforeEach(() => {
            jest.resetModules();
            process.env = {};
        });
        afterAll(() => {
            process.env = OLD_ENV;
        });
        test('throws an error when it encounters an unsupported runtime', () => __awaiter(void 0, void 0, void 0, function* () {
            const lambda = makeMockLambda({
                'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument': {
                    FunctionArn: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
                    Runtime: 'go',
                },
            });
            const cloudWatch = makeMockCloudWatchLogs();
            const settings = {
                flushMetricsToLogs: false,
                layerVersion: 23,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const config = (yield lambda
                .getFunction({ FunctionName: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument' })
                .promise()).Configuration;
            yield expect(function_1.getFunctionConfiguration(lambda, cloudWatch, config, 'us-east-1', settings)).rejects.toThrow();
        }));
        test('replaces the layer arn when the version has changed', () => __awaiter(void 0, void 0, void 0, function* () {
            var _a;
            const lambda = makeMockLambda({
                'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument': {
                    FunctionArn: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
                    Layers: [
                        { Arn: 'arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node12-x:22' },
                        { Arn: 'arn:aws:lambda:us-east-1:464622532012:layer:AnotherLayer:10' },
                    ],
                    Runtime: 'nodejs12.x',
                },
            });
            const cloudWatch = makeMockCloudWatchLogs();
            const settings = {
                flushMetricsToLogs: false,
                layerVersion: 23,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const config = (yield lambda
                .getFunction({ FunctionName: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument' })
                .promise()).Configuration;
            const result = yield function_1.getFunctionConfiguration(lambda, cloudWatch, config, 'us-east-1', settings);
            expect((_a = result.updateRequest) === null || _a === void 0 ? void 0 : _a.Layers).toMatchInlineSnapshot(`
                      Array [
                        "arn:aws:lambda:us-east-1:464622532012:layer:AnotherLayer:10",
                        "arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node12-x:23",
                      ]
                `);
        }));
        test('returns configurations without updateRequest when no changes need to be made', () => __awaiter(void 0, void 0, void 0, function* () {
            const lambda = makeMockLambda({
                'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument': {
                    Environment: {
                        Variables: {
                            [constants_1.FLUSH_TO_LOG_ENV_VAR]: 'false',
                            [constants_1.LAMBDA_HANDLER_ENV_VAR]: 'index.handler',
                            [constants_1.LOG_LEVEL_ENV_VAR]: 'debug',
                            [constants_1.MERGE_XRAY_TRACES_ENV_VAR]: 'false',
                            [constants_1.SITE_ENV_VAR]: 'datadoghq.com',
                            [constants_1.TRACE_ENABLED_ENV_VAR]: 'false',
                        },
                    },
                    FunctionArn: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
                    Handler: '/opt/nodejs/node_modules/datadog-lambda-js/handler.handler',
                    Layers: [{ Arn: 'arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node12-x:22' }],
                    Runtime: 'nodejs12.x',
                },
            });
            const cloudWatch = makeMockCloudWatchLogs();
            const settings = {
                flushMetricsToLogs: false,
                layerVersion: 22,
                logLevel: 'debug',
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const config = (yield lambda
                .getFunction({ FunctionName: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument' })
                .promise()).Configuration;
            expect((yield function_1.getFunctionConfiguration(lambda, cloudWatch, config, 'us-east-1', settings)).updateRequest).toBeUndefined();
        }));
        test('uses the GovCloud lambda layer when a GovCloud region is detected', () => __awaiter(void 0, void 0, void 0, function* () {
            var _b;
            const lambda = makeMockLambda({
                'arn:aws-us-gov:lambda:us-gov-east-1:000000000000:function:autoinstrument': {
                    FunctionArn: 'arn:aws-us-gov:lambda:us-gov-east-1:000000000000:function:autoinstrument',
                    Runtime: 'nodejs12.x',
                },
            });
            const cloudWatch = makeMockCloudWatchLogs();
            const settings = {
                flushMetricsToLogs: false,
                layerVersion: 30,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const config = (yield lambda
                .getFunction({ FunctionName: 'arn:aws-us-gov:lambda:us-gov-east-1:000000000000:function:autoinstrument' })
                .promise()).Configuration;
            const result = yield function_1.getFunctionConfiguration(lambda, cloudWatch, config, 'us-gov-east-1', settings);
            expect((_b = result.updateRequest) === null || _b === void 0 ? void 0 : _b.Layers).toMatchInlineSnapshot(`
                      Array [
                        "arn:aws-us-gov:lambda:us-gov-east-1:002406178527:layer:Datadog-Node12-x:30",
                      ]
                `);
        }));
        test('requests log group configuration when forwarderARN is set', () => __awaiter(void 0, void 0, void 0, function* () {
            ;
            loggroup.calculateLogGroupUpdateRequest.mockImplementation(() => ({ logGroupName: '/aws/lambda/group' }));
            const lambda = makeMockLambda({
                'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument': {
                    FunctionArn: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
                    Handler: 'index.handler',
                    Runtime: 'nodejs12.x',
                },
            });
            const cloudWatch = makeMockCloudWatchLogs();
            const settings = {
                flushMetricsToLogs: false,
                forwarderARN: 'my-forwarder',
                layerVersion: 22,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const config = (yield lambda
                .getFunction({ FunctionName: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument' })
                .promise()).Configuration;
            const result = yield function_1.getFunctionConfiguration(lambda, cloudWatch, config, 'us-east-1', settings);
            expect(result).toBeDefined();
            expect(result.logGroupConfiguration).toMatchInlineSnapshot(`
                Object {
                  "logGroupName": "/aws/lambda/group",
                }
            `);
        }));
    });
    describe('getLambdaConfigsFromRegEx', () => {
        const OLD_ENV = process.env;
        beforeEach(() => {
            jest.resetModules();
            process.env = {};
        });
        afterAll(() => {
            process.env = OLD_ENV;
        });
        test('returns the update request for each function that matches the pattern', () => __awaiter(void 0, void 0, void 0, function* () {
            const lambda = makeMockLambda({
                'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument-scooby': {
                    FunctionArn: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument-scooby',
                    FunctionName: 'autoinstrument-scooby',
                    Handler: 'index.handler',
                    Runtime: 'nodejs12.x',
                },
                'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument-scr.': {
                    FunctionArn: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument-scr.',
                    FunctionName: 'autoinstrument-scr.',
                    Handler: 'index.handler',
                    Runtime: 'nodejs12.x',
                },
            });
            const cloudWatch = makeMockCloudWatchLogs();
            const settings = {
                flushMetricsToLogs: false,
                layerVersion: 22,
                logLevel: 'debug',
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const result = yield function_1.getLambdaConfigsFromRegEx(lambda, cloudWatch, 'us-east-1', 'autoinstrument-scr.', settings);
            expect(result.length).toEqual(1);
            expect(result[0].updateRequest).toMatchInlineSnapshot(`
        Object {
          "Environment": Object {
            "Variables": Object {
              "DD_FLUSH_TO_LOG": "false",
              "DD_LAMBDA_HANDLER": "index.handler",
              "DD_LOG_LEVEL": "debug",
              "DD_MERGE_XRAY_TRACES": "false",
              "DD_SITE": "datadoghq.com",
              "DD_TRACE_ENABLED": "false",
            },
          },
          "FunctionName": "arn:aws:lambda:us-east-1:000000000000:function:autoinstrument-scr.",
          "Handler": "/opt/nodejs/node_modules/datadog-lambda-js/handler.handler",
          "Layers": Array [
            "arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node12-x:22",
          ],
        }
      `);
        }));
        test('fails when retry count is exceeded', () => __awaiter(void 0, void 0, void 0, function* () {
            const makeMockLambdaListFunctionsError = () => ({
                listFunctions: jest.fn().mockImplementation((args) => ({
                    promise: () => Promise.reject(),
                })),
            });
            const lambda = makeMockLambdaListFunctionsError();
            const cloudWatch = makeMockCloudWatchLogs();
            const settings = {
                flushMetricsToLogs: false,
                layerVersion: 22,
                logLevel: 'debug',
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            yield expect(function_1.getLambdaConfigsFromRegEx(lambda, cloudWatch, 'us-east-1', 'fake-pattern', settings)).rejects.toStrictEqual(new Error('Max retry count exceeded.'));
        }));
    });
    describe('updateLambdaConfigs', () => {
        const OLD_ENV = process.env;
        beforeEach(() => {
            jest.resetModules();
            process.env = {};
        });
        afterAll(() => {
            process.env = OLD_ENV;
        });
        test('updates every lambda', () => __awaiter(void 0, void 0, void 0, function* () {
            const lambda = makeMockLambda({});
            const configs = [
                {
                    functionARN: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
                    lambdaConfig: {
                        FunctionArn: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
                        Handler: 'index.handler',
                        Runtime: 'nodejs12.x',
                    },
                    lambdaLibraryLayerArn: 'arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node12-x',
                    updateRequest: {
                        Environment: {
                            Variables: {
                                [constants_1.LAMBDA_HANDLER_ENV_VAR]: 'index.handler',
                                [constants_1.MERGE_XRAY_TRACES_ENV_VAR]: 'false',
                                [constants_1.TRACE_ENABLED_ENV_VAR]: 'false',
                            },
                        },
                        FunctionName: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
                        Handler: '/opt/nodejs/node_modules/datadog-lambda-js/handler.handler',
                        Layers: ['arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node12-x:22'],
                    },
                },
            ];
            const cloudWatch = makeMockCloudWatchLogs();
            yield function_1.updateLambdaConfigs(lambda, cloudWatch, configs);
            expect(lambda.updateFunctionConfiguration).toHaveBeenCalledWith({
                Environment: {
                    Variables: {
                        [constants_1.LAMBDA_HANDLER_ENV_VAR]: 'index.handler',
                        [constants_1.MERGE_XRAY_TRACES_ENV_VAR]: 'false',
                        [constants_1.TRACE_ENABLED_ENV_VAR]: 'false',
                    },
                },
                FunctionName: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
                Handler: '/opt/nodejs/node_modules/datadog-lambda-js/handler.handler',
                Layers: ['arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node12-x:22'],
            });
        }));
    });
    describe('getLayerArn', () => {
        const OLD_ENV = process.env;
        beforeEach(() => {
            jest.resetModules();
            process.env = {};
        });
        afterAll(() => {
            process.env = OLD_ENV;
        });
        test('gets sa-east-1 Node12 Lambda Library layer ARN', () => __awaiter(void 0, void 0, void 0, function* () {
            const runtime = 'nodejs12.x';
            const settings = {
                flushMetricsToLogs: false,
                layerAWSAccount: mockAwsAccount,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const region = 'sa-east-1';
            const layerArn = function_1.getLayerArn(runtime, settings, region);
            expect(layerArn).toEqual(`arn:aws:lambda:${region}:${mockAwsAccount}:layer:Datadog-Node12-x`);
        }));
        test('gets sa-east-1 Python37 gov cloud Lambda Library layer ARN', () => __awaiter(void 0, void 0, void 0, function* () {
            const runtime = 'python3.7';
            const settings = {
                flushMetricsToLogs: false,
                layerAWSAccount: mockAwsAccount,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const region = 'us-gov-1';
            const layerArn = function_1.getLayerArn(runtime, settings, region);
            expect(layerArn).toEqual(`arn:aws-us-gov:lambda:${region}:${constants_1.GOVCLOUD_LAYER_AWS_ACCOUNT}:layer:Datadog-Python37`);
        }));
    });
    describe('getExtensionArn', () => {
        const OLD_ENV = process.env;
        beforeEach(() => {
            jest.resetModules();
            process.env = {};
        });
        afterAll(() => {
            process.env = OLD_ENV;
        });
        test('gets sa-east-1 Lambda Extension layer ARN', () => __awaiter(void 0, void 0, void 0, function* () {
            const settings = {
                flushMetricsToLogs: false,
                layerAWSAccount: mockAwsAccount,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const region = 'sa-east-1';
            const layerArn = function_1.getExtensionArn(settings, region);
            expect(layerArn).toEqual(`arn:aws:lambda:${region}:${mockAwsAccount}:layer:Datadog-Extension`);
        }));
        test('gets sa-east-1 gov cloud Lambda Extension layer ARN', () => __awaiter(void 0, void 0, void 0, function* () {
            const settings = {
                flushMetricsToLogs: false,
                layerAWSAccount: mockAwsAccount,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const region = 'us-gov-1';
            const layerArn = function_1.getExtensionArn(settings, region);
            expect(layerArn).toEqual(`arn:aws-us-gov:lambda:${region}:${constants_1.GOVCLOUD_LAYER_AWS_ACCOUNT}:layer:Datadog-Extension`);
        }));
    });
    describe('calculateUpdateRequest', () => {
        const OLD_ENV = process.env;
        beforeEach(() => {
            jest.resetModules();
            process.env = {};
        });
        afterAll(() => {
            process.env = OLD_ENV;
        });
        test('calculates an update request with just lambda library layers', () => {
            const config = {
                FunctionArn: 'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world',
                Handler: 'index.handler',
                Layers: [],
            };
            const settings = {
                flushMetricsToLogs: false,
                layerAWSAccount: mockAwsAccount,
                layerVersion: 5,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const lambdaLibraryLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Node12-x`;
            const lambdaExtensionLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Extension`;
            const runtime = 'nodejs12.x';
            const updateRequest = function_1.calculateUpdateRequest(config, settings, lambdaLibraryLayerArn, lambdaExtensionLayerArn, runtime);
            expect(updateRequest).toMatchInlineSnapshot(`
        Object {
          "Environment": Object {
            "Variables": Object {
              "DD_FLUSH_TO_LOG": "false",
              "DD_LAMBDA_HANDLER": "index.handler",
              "DD_MERGE_XRAY_TRACES": "false",
              "DD_SITE": "datadoghq.com",
              "DD_TRACE_ENABLED": "false",
            },
          },
          "FunctionName": "arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world",
          "Handler": "/opt/nodejs/node_modules/datadog-lambda-js/handler.handler",
          "Layers": Array [
            "arn:aws:lambda:sa-east-1:123456789012:layer:Datadog-Node12-x:5",
          ],
        }
      `);
        });
        test('calculates an update request with a lambda library, extension, and DATADOG_API_KEY', () => {
            process.env.DATADOG_API_KEY = '1234';
            const config = {
                FunctionArn: 'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world',
                Handler: 'index.handler',
                Layers: [],
            };
            const settings = {
                extensionVersion: 6,
                flushMetricsToLogs: false,
                layerAWSAccount: mockAwsAccount,
                layerVersion: 5,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const lambdaLibraryLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Node12-x`;
            const lambdaExtensionLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Extension`;
            const runtime = 'nodejs12.x';
            const updateRequest = function_1.calculateUpdateRequest(config, settings, lambdaLibraryLayerArn, lambdaExtensionLayerArn, runtime);
            expect(updateRequest).toMatchInlineSnapshot(`
        Object {
          "Environment": Object {
            "Variables": Object {
              "DD_API_KEY": "1234",
              "DD_FLUSH_TO_LOG": "false",
              "DD_LAMBDA_HANDLER": "index.handler",
              "DD_MERGE_XRAY_TRACES": "false",
              "DD_SITE": "datadoghq.com",
              "DD_TRACE_ENABLED": "false",
            },
          },
          "FunctionName": "arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world",
          "Handler": "/opt/nodejs/node_modules/datadog-lambda-js/handler.handler",
          "Layers": Array [
            "arn:aws:lambda:sa-east-1:123456789012:layer:Datadog-Extension:6",
            "arn:aws:lambda:sa-east-1:123456789012:layer:Datadog-Node12-x:5",
          ],
        }
      `);
        });
        test('calculates an update request with a lambda library, extension, and DATADOG_KMS_API_KEY', () => {
            process.env.DATADOG_KMS_API_KEY = '5678';
            const config = {
                FunctionArn: 'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world',
                Handler: 'index.handler',
                Layers: [],
            };
            const settings = {
                extensionVersion: 6,
                flushMetricsToLogs: false,
                layerAWSAccount: mockAwsAccount,
                layerVersion: 5,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const lambdaLibraryLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Python36`;
            const lambdaExtensionLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Extension`;
            const runtime = 'python3.6';
            const updateRequest = function_1.calculateUpdateRequest(config, settings, lambdaLibraryLayerArn, lambdaExtensionLayerArn, runtime);
            expect(updateRequest).toMatchInlineSnapshot(`
        Object {
          "Environment": Object {
            "Variables": Object {
              "DD_FLUSH_TO_LOG": "false",
              "DD_KMS_API_KEY": "5678",
              "DD_LAMBDA_HANDLER": "index.handler",
              "DD_MERGE_XRAY_TRACES": "false",
              "DD_SITE": "datadoghq.com",
              "DD_TRACE_ENABLED": "false",
            },
          },
          "FunctionName": "arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world",
          "Handler": "datadog_lambda.handler.handler",
          "Layers": Array [
            "arn:aws:lambda:sa-east-1:123456789012:layer:Datadog-Extension:6",
            "arn:aws:lambda:sa-east-1:123456789012:layer:Datadog-Python36:5",
          ],
        }
      `);
        });
        test('by default calculates an update request with DATADOG_SITE being set to datadoghq.com', () => {
            const config = {
                FunctionArn: 'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world',
                Handler: 'index.handler',
                Layers: [],
            };
            const settings = {
                flushMetricsToLogs: false,
                layerAWSAccount: mockAwsAccount,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const lambdaLibraryLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Python36`;
            const lambdaExtensionLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Extension`;
            const runtime = 'python3.6';
            const updateRequest = function_1.calculateUpdateRequest(config, settings, lambdaLibraryLayerArn, lambdaExtensionLayerArn, runtime);
            expect(updateRequest).toMatchInlineSnapshot(`
        Object {
          "Environment": Object {
            "Variables": Object {
              "DD_FLUSH_TO_LOG": "false",
              "DD_LAMBDA_HANDLER": "index.handler",
              "DD_MERGE_XRAY_TRACES": "false",
              "DD_SITE": "datadoghq.com",
              "DD_TRACE_ENABLED": "false",
            },
          },
          "FunctionName": "arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world",
          "Handler": "datadog_lambda.handler.handler",
        }
      `);
        });
        test('calculates an update request with DATADOG_SITE being set to datadoghq.eu', () => {
            process.env.DATADOG_SITE = 'datadoghq.eu';
            const config = {
                FunctionArn: 'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world',
                Handler: 'index.handler',
                Layers: [],
            };
            const settings = {
                flushMetricsToLogs: false,
                layerAWSAccount: mockAwsAccount,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const lambdaLibraryLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Python36`;
            const lambdaExtensionLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Extension`;
            const runtime = 'python3.6';
            const updateRequest = function_1.calculateUpdateRequest(config, settings, lambdaLibraryLayerArn, lambdaExtensionLayerArn, runtime);
            expect(updateRequest).toMatchInlineSnapshot(`
        Object {
          "Environment": Object {
            "Variables": Object {
              "DD_FLUSH_TO_LOG": "false",
              "DD_LAMBDA_HANDLER": "index.handler",
              "DD_MERGE_XRAY_TRACES": "false",
              "DD_SITE": "datadoghq.eu",
              "DD_TRACE_ENABLED": "false",
            },
          },
          "FunctionName": "arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world",
          "Handler": "datadog_lambda.handler.handler",
        }
      `);
        });
        test('throws an error when an invalid DATADOG_SITE url is given', () => {
            process.env.DATADOG_SITE = 'datacathq.eu';
            const config = {
                FunctionArn: 'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world',
                Handler: 'index.handler',
                Layers: [],
            };
            const settings = {
                flushMetricsToLogs: false,
                layerAWSAccount: mockAwsAccount,
                layerVersion: 5,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const lambdaLibraryLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Python36`;
            const lambdaExtensionLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Extension`;
            const runtime = 'python3.6';
            expect(() => {
                function_1.calculateUpdateRequest(config, settings, lambdaLibraryLayerArn, lambdaExtensionLayerArn, runtime);
            }).toThrowError('Warning: Invalid site URL. Must be either datadoghq.com, datadoghq.eu, us3.datadoghq.com, or ddog-gov.com.');
        });
        test('throws an error when neither DATADOG_API_KEY nor DATADOG_KMS_API_KEY are given through the environment while using extensionVersion', () => {
            const config = {
                FunctionArn: 'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world',
                Handler: 'index.handler',
                Layers: [],
            };
            const settings = {
                extensionVersion: 6,
                flushMetricsToLogs: false,
                layerAWSAccount: mockAwsAccount,
                layerVersion: 5,
                mergeXrayTraces: false,
                tracingEnabled: false,
            };
            const lambdaLibraryLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Python36`;
            const lambdaExtensionLayerArn = `arn:aws:lambda:sa-east-1:${mockAwsAccount}:layer:Datadog-Extension`;
            const runtime = 'python3.6';
            expect(() => {
                function_1.calculateUpdateRequest(config, settings, lambdaLibraryLayerArn, lambdaExtensionLayerArn, runtime);
            }).toThrowError("When 'extensionLayer' is set, DATADOG_API_KEY or DATADOG_KMS_API_KEY must also be set");
        });
    });
});
