// @datadog/datadog-ci/dist/commands/synthetics/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExecutionRule = exports.Operator = exports.ERRORS = void 0;
var ERRORS;
(function (ERRORS) {
    ERRORS["TIMEOUT"] = "Timeout";
    ERRORS["ENDPOINT"] = "Endpoint Failure";
    ERRORS["TUNNEL"] = "Tunnel Failure";
})(ERRORS = exports.ERRORS || (exports.ERRORS = {}));
var Operator;
(function (Operator) {
    Operator["contains"] = "contains";
    Operator["doesNotContain"] = "doesNotContain";
    Operator["is"] = "is";
    Operator["isNot"] = "isNot";
    Operator["lessThan"] = "lessThan";
    Operator["matches"] = "matches";
    Operator["doesNotMatch"] = "doesNotMatch";
    Operator["validates"] = "validates";
    Operator["isInMoreThan"] = "isInMoreThan";
    Operator["isInLessThan"] = "isInLessThan";
})(Operator = exports.Operator || (exports.Operator = {}));
var ExecutionRule;
(function (ExecutionRule) {
    ExecutionRule["BLOCKING"] = "blocking";
    ExecutionRule["NON_BLOCKING"] = "non_blocking";
    ExecutionRule["SKIPPED"] = "skipped";
})(ExecutionRule = exports.ExecutionRule || (exports.ExecutionRule = {}));
